contracts/adapters/TargetWeightWrapExtension.sol

```diff
@@ -0,0 +1,382 @@
+        string wrapAdapterName;      // Name of the wrap adapter to use
+        bytes wrapData;              // Data for wrapping
+        bytes unwrapData;            // Data for unwrapping
+    }
+
+    /* ============ Events ============ */
+
+    event RebalanceAccessUpdated(bool isRebalanceOpen);
+    event TargetsSet(
+        address indexed reserveAsset,
+        uint256 minReserveWeight,
+        uint256 maxReserveWeight,
+        address[] targetAssets,
+        TargetWeightWrapParams[] executionParams
+    );
+    event RebalancePaused();
+
+    /* ========== Immutables ========= */
+
+    ISetToken public immutable setToken;
+    IWrapModuleV2 public immutable wrapModule;
+    ISetValuer public immutable setValuer;
+
+    /* ========== State Variables ========= */
+
+    bool public isRebalancingActive;
+    bool public isRebalanceOpen;
+
+    mapping(address => TargetWeightWrapParams) public executionParams; 
+    RebalanceInfo public rebalanceInfo;
+
+    /* ============ Modifiers ============ */
+
+    modifier onlyAllowedRebalancer() {
+        _validateOnlyAllowedRebalancer();
+        _;
+    }
+
+    /* ============ Constructor ============ */
+
+    /**
+     * @notice Initializes the extension with the required contracts and parameters.
+     * @param _manager Address of Index Manager contract
+     * @param _wrapModule Address of IWrapModuleV2 for wrapping and unwrapping reserve asset
+     * @param _setValuer Address of SetValuer for calculating valuations and weights
+     * @param _isRebalanceOpen Flag indicating if anyone can rebalance
+     */
+    constructor(
+        IBaseManager _manager,
+        IWrapModuleV2 _wrapModule,
+        ISetValuer _setValuer,
+        bool _isRebalanceOpen
+    ) public BaseExtension(_manager) {
+        manager = _manager;
+        setToken = manager.setToken();
+        wrapModule = _wrapModule;
+        setValuer = _setValuer;
+        isRebalanceOpen = _isRebalanceOpen;
+    }
+
+    /* ========== Rebalance Functions ========== */
+
+    /**
+     * @notice Wraps reserve asset units into the target asset. 
+     * @dev Must be called when the reserve asset is overweight, rebalancing is enabled, and the caller is an allowed rebalancer. 
+     * Ensures that after wrapping, the target asset does not become overweight and the reserve asset does not become underweight.
+     * @param _targetAsset Address of the target asset to wrap into.
+     * @param _reserveUnits Units of the reserve asset to wrap.
+     */
+    function wrap(
+        address _targetAsset,
+        uint256 _reserveUnits
+    )
+        external
+        nonReentrant
+        onlyAllowedRebalancer
+    {
+        require(isRebalancingActive, "Rebalancing is not active");
+        require(rebalanceInfo.targetAssets.contains(_targetAsset), "Invalid target asset");
+        require(isReserveOverweight(), "Reserve asset is not overweight");
```

Comment:
Wouldn't it be better for to allow wraps so long as the reserve is _not underweight_? 

That way rebalancers can maximize performance be keeping less of the value in the non-yielding reserve asset.

If we require the reserve asset to be overweight, there may be situations where a yielding asset is under its target weight, but we can't rebalance into it.

Reply:
yes this makes sense...in this case it may make sense to remove the pre-wrap weight check entirely

Reply:
removed condition on `wrap()` here [`4269d1e`](https://github.com/IndexCoop/index-coop-smart-contracts/pull/182/commits/4269d1ee222d2de41f3a5572ab5e0b2fa28aeda7)